### PR TITLE
Sticky session

### PR DESCRIPTION
**Pull request by BlackYoup**
This is a WIP pull request to support sticky sessions ( see #94 )

Right now, the sticky session cookie is hardcoded everywhere, I plan to make it configurable.
In this initial pull request, there is:
- partial support for parsing cookies (the parser take_while(is_alphanumeric) and it's not great). Also, I don't handle yet cookies larger than the input buffer (if I understood correctly)
- check if clients has a sticky session cookie and store it in the Connection
- remove Set-Cookie header from backend with the value of the sticky session cookie name
- untested "connect to backend if sticky session cookie matches a backend"

It still misses:
- Set the sticky cookie if not yet set
- Store the backend related to the cookie
- maybe other things

### CODE REVIEW

**Comment by BlackYoup** [comment]
Alright, so it's still a bit raw but it works™

Every backend now has a `pub id: Uuid` field.

How it works:
- the client sends a request
- the `parse_request` method checks if the client has the `sticky_session` cookie and stores the value of the cookie if there is one. It removes the cookie from the buffer if it is found.
- when we want to connect to the backend, we check if the request backend exists by looping against all backends available for the application. If it is found, it is returned. If it is not, we default to the regular connection process
- the backend answers, the `parse_response` checks if there is a `Set-Cookie` header with the same cookie name as our `sticky_session` cookie name.  If there is, we remove it (so the backend can't set a sticky session cookie on its own).
- the `parse_response` then check if we need to set a sticky_session cookie. It does that by:
1) checking if the application has the `sticky_session` cookie option
2) checking if the client sent a `sticky_session` cookie: if it doesn't, then we set it.
3) checking if the value of the `sticky_session` cookie sent by the client is the same as the backend id used to handle the request. If it is not, we send the cookie with the new backend id.

I'll try to squash some commits and re-arrange them but I think that's not gonna be simple.

The variable / struct / fields naming can sometimes be a bit hard to follow here and there, maybe there are better names. 

No tests added for now, I'll add them once you validated all that.

**Comment by BlackYoup** [comment]
There is still a small bug in the cookie parsing.. Will fix it later. You can test with multiple cookies like: `curl -H "Cookie:TEST=BLAH;SOZUBALANCEID=bdb15261-e801-4558-8476-43b52de5bfeb; SWAG=MEH"`

It won't work if there is only one cookie, the proxy will loop indefinitely.

File: ctl/src/command.rs [hunk 1]
```diff
@@ -478,7 +479,8 @@ pub fn add_frontend(channel: &mut Channel<ConfigMessage,ConfigMessageAnswer>, ap
               app_id: String::from(app_id),
               hostname: String::from(hostname),
               path_begin: String::from(path_begin),
-              fingerprint: CertFingerprint(fingerprint)
+              fingerprint: CertFingerprint(fingerprint),
+              sticky_session: bool::from_str(sticky_session).expect("sticky_session should be a boolean")
```
Comment (Geal) [inline ctl/src/command.rs]: this seems weird, the boolean should be parsed much earlier

File: ctl/src/main.rs [hunk 2]
```diff
@@ -233,18 +233,20 @@ fn main() {
...
-          let id          = frontend_sub.value_of("id").expect("missing id");
-          let hostname    = frontend_sub.value_of("hostname").expect("missing frontend hostname");
-          let path_begin  = frontend_sub.value_of("path_begin").unwrap_or("");
-          let certificate = frontend_sub.value_of("certificate");
-          remove_frontend(&mut channel, id, hostname, path_begin, certificate);
+          let id              = frontend_sub.value_of("id").expect("missing id");
+          let hostname        = frontend_sub.value_of("hostname").expect("missing frontend hostname");
+          let path_begin      = frontend_sub.value_of("path_begin").unwrap_or("");
+          let certificate     = frontend_sub.value_of("certificate");
+          let sticky_session  = frontend_sub.value_of("sticky_session").unwrap_or("false");
```
Comment (Geal) [inline ctl/src/main.rs]: the `sticky_session` parameter should be converted to boolean here

File: lib/src/network/http.rs [hunk 3]
```diff
@@ -523,8 +558,19 @@ impl ProxyConfiguration<Client> for ServerConfiguration {
     };
 
     //FIXME: too many unwraps here
+    let sticky_session = {
+      if let Some(ref request_state) = client.http().unwrap().state.as_ref().unwrap().request {
```
Comment (Geal) [inline lib/src/network/http.rs]: this could be a method of the state

File: lib/src/network/mod.rs [hunk 4]
```diff
@@ -143,6 +144,7 @@ pub enum BackendStatus {
 
 #[derive(Debug,PartialEq,Eq)]
 pub struct Backend {
+  pub id:                 Uuid,
```
Comment (Geal) [inline lib/src/network/mod.rs]: why a UUID? This could be just a regular integer, no?

Comment (BlackYoup) [inline lib/src/network/mod.rs]: It could. By regular integer do you mean an increasing integer ? It just felt like easier with UUID

Comment (Geal) [inline lib/src/network/mod.rs]: it's just a question of cost, but I don't think it's really significant. a UUID is 128 bits, the incrementing integer could just be 32 bits. From what I see, there's one id per backend, and an option of id per client?

Comment (BlackYoup) [inline lib/src/network/mod.rs]: > From what I see, there's one id per backend, and an option of id per client?

yes, it is.

I agree on the cost it adds, I'll make the change.

**Comment by BlackYoup** [comment]
This PR now depends on https://github.com/sozu-proxy/sozu/pull/221 to properly parse request cookies

You might want to have a look at the last commit, it changes how we remove the sticky cookie from the request cookies. I couldn't rebase it without editing multiple commits at once
